### PR TITLE
Open STRICT_STYLE=false to account for clients using restricted characters [GEOS-6811]

### PR DESCRIPTION
**Pull request by jodygarnett**
Two part fix:

* Change Paths to allow some invalid characters with a warning. This allows GeoServer 2.5.x data directories to load (for manual fix)
* Introduce -DSTRICT_STYLE=FALSE allowing GetMap and GetLegendGraphic lookup to try replacing invalid characters with _ (to account for styles that hae been fixed)

### CODE REVIEW

File: src/main/src/main/java/org/geoserver/catalog/impl/CatalogImpl.java [hunk 1]
```diff
@@ -1280,15 +1280,28 @@ public StyleInfo getStyle(String id) {
...
+                String resource2 = resource.replaceAll(REPLACE, "_");
+                String name2 = name.replaceAll(REPLACE, "_");
+                // check workspace: sty_le
+                result = getStyleByName(prefix, resource2);
+                // check workspace_sty_le
+                result = facade.getStyleByName(name2);
+                // check workspace: sty:le
+                result = getStyleByName(prefix, resource);
+                // check workspace:sty:le
+                result = facade.getStyleByName(name);
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/catalog/impl/CatalogImpl.java]: This order looks correct.

File: src/main/src/main/java/org/geoserver/catalog/impl/CatalogImpl.java [hunk 2]
```diff
@@ -1280,15 +1280,28 @@ public StyleInfo getStyle(String id) {
         return facade.getStyle(id);
     }
 
+    static final String REPLACE = "[:]";
+    static final boolean STRICT_STYLE = Boolean.valueOf(System.getProperty("STRICT_STYLE", "true"));
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/catalog/impl/CatalogImpl.java]: May wish to mark this non final to aid in testing.

**Comment by jodygarnett** [comment]
@travislbrundage see above comments.

**Comment by aaime** [comment]
Is the replacement of "weird" chars with _ already in place?

**Comment by jodygarnett** [comment]
No, currently weird characters just fail. Advice will be to manually update.

**Comment by aaime** [comment]
Confused... by your description it seems this fix works around an issue that cannot be generated, because there is not char substitution?

**Comment by travislbrundage** [comment]
@aaime I'll attempt to explain.

Basically, if we get a request for some style, sty_le, we'll check twice. Once we'll check literally for the style named sty_le in the catalog, but if that's not found, we'll check for "weird chars" as a substitution. For now it's just sty:le.

The tricky part here is since we are handling colons, it might be workspace:style, or just a style name with a colon in it.

The goal is just to make this regression in names easier to deal with and give the user some leeway until they manually update to use valid style names.

I also noticed this code is actually wrong. It needs to check if it finds the result before doing another attempt. I'll update with a patch shortly.

**Comment by aaime** [comment]
Travis, your explation helps some, but I still do not know, what put the _ in the names in the first place? 
I guess at some point (or even currently) we had code that transformed weird chars in _ during catalog insertion, and so we have to lookup for the transformed name, or something like that.
Where is the code that this this transformation?

**Comment by jodygarnett** [comment]
I am going to catch up with travis on this one tomorrow. My initial proposal to gather up name checking in a utility class got stuck on this issue of how to "automatically update". The idea of "fixing" on load was rejected - so this pull request is going for a manual process.

Andrea would it be sufficient to:

1) add warnings for administrator to correct (much like is done for security warnings at present).
2) have a "fix me" link next to the invalid name that performs introduces underscores by default

It will be ironic if this discussion reintroduces the ResourceInfo Alias (as a way to smoothly point "foo::bar" alias to the replacement "foo__bar". It would have the advantage of being able to handle name collisions - so "foo::bar" could point to "foo__bar3".

**Comment by aaime** [comment]
Jody, I'm lost, can we start from the beginning? What is the problem that you're trying to solve here? The pull request seems to address cases in which the _ was already introduced, but afaik GeoServer never did that.

**Comment by jodygarnett** [comment]
We are trying to solve a regression in the catalog introduced by GeoServer 2.6. Filenames for styles that used to be valid (example "foo::bar") are no longer valid due to the file name checking performed by ResourceStore and friends.

Fixing this issue is in three parts:

1. Ensure the GeoServer 2.5.x catalog can load (even with warnings)
2. Provide a mechanism to fix the file names (either manual or automatic)
3. Allow clients that were developed with the invalid names to continue to function (while they transition over to the new names)

This pull request glosses over step number 2.

**Comment by jodygarnett** [comment]
This pull request has been replaced by #960 after much discussion.

**Pull request closed without merging**